### ISSUE TITLE
prefer-arrow-callback should allow named function that isn't being referenced

### ISSUE THREAD

**Issue by vtfn** [eslint/eslint#5675]
Using a named callback improve debuggin by making stack traces more descritive, so prefer-arrow-callback should not warn if a named function is being passed as a callback, even if it isn't been referenced.

```js
function foo(fn) {
    fn();
}

foo(function named() {
    console.log('this is just a useless function example');
});
```

Trying to give as much info as possible:

The version of ESLint I'm using is 2.4.0
The code is above.
The output is: 5:5  error  Unexpected function expression  prefer-arrow-callback
And I'm just setting "prefer-arrow-callback": 2 and "id-length": [2, {"min": 2}]

If any more info is needed please tell me, I don't know what else to include here.

**Comment by eslintbot** [comment]
Hi @vtfn, thanks for the issue. It looks like there's not enough information for us to know how to help you. If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by alberto** [comment]
So what are you proposing? An option to `prefer-arrow-callback` so that it only applies to anonymous functions?

**Comment by vtfn** [comment]
Actualy I think it should only apply to anonymous functions by default since one cannot have a named arrow function. But an option could be ok as well.

**Comment by platinumazure** [comment]
@vtfn If it applied by default, that would be a breaking change and we would have to wait for the next major release. I would also argue that the rule name does not lend itself to an assumption that named functions should be allowed, so I would personally consider that behavior a weird default. I'm sure we can do an enhancement where it is an option people could use, though.

**Comment by vtfn** [comment]
I understand,  well an option should do it just fine then, and I agree with the name argument you've raised

**Comment by platinumazure** [comment]
@eslint/eslint-team What do you think? Should we accept this?

**Comment by xjamundx** [comment]
Sounds to me like a diff rule altogether. Either you want named function
expressions or you want arrow functions. I don't see the use case for
"either".

**Comment by platinumazure** [comment]
@xjamundx Iterating on your thought for a moment, maybe this rule should be
deprecated in favor of a "callback-style" rule? User would choose between
named function expression, anonymous function expression, or arrow function
(or any combination thereof). What do you think?

**Comment by xjamundx** [comment]
@platinumazure Yeah, I think that makes more sense than adding an option here.

**Comment by nzakas** [comment]
We have to be very careful about creating new rules at this point. We basically can't remove rules anymore, because doing so breaks existing configs, so adding a new rule that does more or less the same thing means we will need to keep around two similar rules. Is this important enough to do that?

**Comment by alberto** [comment]
@xjamundx I see why someone could want to use arrow functions for simple one-liners like `filter(x => x % 2)` but use named function expressions for more complex ones, to aid debugging.

**Comment by vtfn** [comment]
@alberto That is exactly my point. So I think it's valid to have an option to ignore `prefer-arrow-callback` if the callback function has a name.

**Comment by nzakas** [comment]
I think an option for this is fine, can anyone suggest an option name?

Also, does anyone want to champion this change?

**Comment by platinumazure** [comment]
@nzakas Maybe `allowNamedFunctions: true` (false by default)?

Complete sample schema:

```json
{
    "rules": {
        " prefer-arrow-callback": ["error", { "allowNamedFunctions": true }]
    }
}
```

**Comment by vtfn** [comment]
how about  a boolean `ignore-named`?

**Comment by alberto** [comment]
@nzakas I do.

+1 to `allowNamedFunctions`

**Comment by nzakas** [comment]
Okay, @alberto is championing this change. Please work with @vtfn to get this implemented.

Also, :+1: to `allowNamedFunctions`

**Comment by alberto** [comment]
@vtfn are you interested in implementing this yourself?

**Comment by vtfn** [comment]
@alberto  I'd love to, but since I've never done a change to help in this project, I'd have to study the source code first and I'm too busy to do it right now. If it isn't too much problem for you, could you do this one? Or, I can wait until I can do it myself.

**Comment by lemonmade** [comment]
I'm interested in starting to contribute to ESLint, if no one else wants to I can take a stab at this.

**Comment by alberto** [comment]
Sorry @lemonmade, I have just pushed this.

**Comment by lemonmade** [comment]
No worries!
